Restructured API

Elix-x opened the pull request:
Completely restructured API splitting it into modules.
Not everything was moved, but it will be fixed & finished in close future.

JoJoDeveloping committed:
Removed all registries and definition classes from the API\n Internal code uses their internal implementation until there is a rework

Elix-x committed:
Created modules API entry point

Created modules API entry point.

Elix-x committed:
Reorganized API - part 1

Reorganized API - part 1: created package structure, moved most of the
classes.

Elix-x committed:
APIs for external modules

APIs for external modules.

Elix-x committed:
Checked module names

Elix-x committed:
Reworked definitons structure

Reworked definitons structure.

Elix-x committed:
Fixed return type

Elix-x committed:
Properties for sub definitions

Properties for sub definitions.

Elix-x committed:
Implemented new API on core module

Implemented new API on core module.
Provided a few examples.

Elix-x committed:
Moved back crystal seed

Elix-x committed:
Implemented new API on all modules

Implemented new API on all modules.
Actual definitions still have to be moved over.

Elix-x committed:
New IMC system

New IMC messaging system.
Has support for sending any object instance.

Elix-x committed:
More methods for external modules

Added more methods for external modules, for handling events from AE and
forge.
Methods are in order of events being called (relative) - first from AE,
then from Forge (may not be the case for IMC).

Elix-x committed:
Added null check in case module is not loaded

JoJoDeveloping committed:
I'm giving up. Too many bugs. Someone please fix this..

Elix-x committed:
Reparametrized tile defs

Elix-x committed:
Generalized definition builders

Elix-x committed:
Further improvements to the def builders

Elix-x committed:
Better type params for TEs

Elix-x committed:
Use resource location as identifier

Elix-x committed:
Type params overload

Elix-x committed:
Builder to rule them all

Elix-x committed:
Removed damaged item def

Will be replaced with subdefinitions

Elix-x committed:
Definition implementation to rule them all

Sub definitions - TODO.

Elix-x committed:
Reimplemented base definitions

Elix-x committed:
Corrected builder return types

Elix-x committed:
Reimplemented definition builders

Elix-x committed:
Created definitions provider

In order to "connect" modules.

Elix-x committed:
Removed item factory

Elix-x committed:
Fixed forgotten definition identifier types

Elix-x committed:
Moved definitions provider

Elix-x committed:
Created current module

AKA module which currently handles event.

Elix-x committed:
Linked definitions

Block -> Item
Tile -> Block

Elix-x committed:
Added default string id method

Might be removed in the future.

Elix-x committed:
Fixed visibility

Elix-x committed:
Fixed registration order

Elix-x committed:
Formatted code

Elix-x committed:
Implemented and provided example for crank

Elix-x committed:
Fixed definitions loading order in other modules

Elix-x committed:
Got rid of suppliers

Unnecessary useless everything size eating stuff.

Elix-x committed:
Implemented sub definitions basis

Elix-x committed:
Make sure that property is valid

Elix-x committed:
More generics for sub defs & props

Elix-x committed:
Properties have names

Elix-x committed:
Fixed visibility modifiers

Elix-x committed:
Generics fixes

Elix-x committed:
Hooked subdefs providers into def builders

Elix-x committed:
Better subdef generics for easier implementation

Elix-x committed:
Reorganized way of setting default subdef

Elix-x committed:
Revert "Better subdef generics for easier implementation"

This reverts commit 8a7dce65a3555cd40f1bee0b2b93bb68007b22b6.

Elix-x committed:
Sub definitions are auto generated

They cannot be set manually.
Def builders will automatically generate subdefinitions structure if
needed.

Elix-x committed:
Implemented subdefinitions for blocks

Using existing blockstate & property structure.

Elix-x committed:
Created IStateItem

Interface for easier use of items with states.
Equivalent to IBlockState structure.

Elix-x committed:
Implemented subdefinitions for items

Using IStateItem structure created earlier.

Elix-x committed:
Changed state to a class

With all methods implemented.

Elix-x committed:
Use state instead of itemstack for sub definitions

For many reasons.
Also, added toItemStack to State method for shortcutting.

Elix-x committed:
Removed colored item definition

Will be repleaced with sub definitions.

Elix-x committed:
Some additional interface refactoring

Elix-x committed:
Provided example of IStateItem implementation

Provided example of IStateItem implementation for paint ball.

JoJoDeveloping committed:
Added definitions for core module
Also added createDefaultItemBlock()

Elix-x committed:
Changed callbacks to use def, added build callback

FML lifecycle callbacks now get definition instead of instance.
Build callbacks are called right after definition is built.

Elix-x committed:
Implemented new way of creating default item blocks

Using new build callback.

Elix-x committed:
Item defs for blocks

Elix-x committed:
Wrapping up - pass 1

Registries no longer are represented by interfaces. They may come back
in the future.
As is API and generic def holders. Implementations are deprecated and to
be moved.

Elix-x committed:
Wrapping up - pass 2 - fixed most of compile errors

A few are still left though.

JoJoDeveloping committed:
Reworked the ItemBlock registration dispatching

JoJoDeveloping committed:
Added all block items

JoJoDeveloping committed:
Trying to generify, attempt #1

JoJoDeveloping committed:
Generified definitions

JoJoDeveloping committed:
Definitions now have a special IDefinition sub-implementation

JoJoDeveloping committed:
Revert "Definitions now have a special IDefinition sub-implementation"

This reverts commit f4990ae0de7da0405631fc0298a79cbd86d02af7.

JoJoDeveloping committed:
Definitions now use I<X>Defintion if needed

Elix-x committed:
Revert "Definitions now use I<X>Defintion if needed"

This reverts commit f3ece74216797b983664366f9d039f1e07a437c4.

Elix-x committed:
Advanced type parameters for definitions

Implemented more advanced type parameters for definitions - defining
type of definition.
Replaces f3ece74216797b983664366f9d039f1e07a437c4.

Elix-x committed:
Added test code for defs type params checking

Which should always compile properly without any additional casts, with
any type param changes to definitions structure.
TODO: For removal once definitions are stable.

Elix-x committed:
Fixed reinstantiation of mod-modules

If mod is a module at the same time, we need to make sure we use the
same class instance as module, as one used as mod.

Elix-x committed:
Biome & Dimension definition fixes

Fixed compile errors in spatial module biome & dimension type defs.

Elix-x committed:
Merge remote-tracking branch 'refs/remotes/origin/1.10.2-rv3-restructuring' into 1.10.2-rv3-api-boom

Resolved PR merge conflicts.
# Conflicts:
#	src/api/java/appeng/api/features/IRegistryContainer.java
#	src/api/java/appeng/core/spatial/api/movable/IMovableHandler.java
#	src/api/java/appeng/core/spatial/api/movable/IMovableTile.java
#	src/main/java/appeng/core/Registration.java
#	src/main/java/appeng/core/lib/api/imc/IMCBlackListSpatial.java
#	src/main/java/appeng/core/lib/api/imc/IMCSpatial.java

#	src/main/java/appeng/core/lib/features/registries/MovableTileRegistry.java

#	src/main/java/appeng/core/lib/features/registries/RegistryContainer.java
#	src/main/java/appeng/core/spatial/DefaultSpatialHandler.java
#	src/main/java/appeng/core/spatial/world/CachedPlane.java

Elix-x commented:
If anybody has any questions/concerns - ask them now.

Elix-x committed:
Fixed more compile errors

Fixed more compile errors in AppEngApi and some in ApiParts.

Elix-x committed:
Moved item definitions

Moved item definitions to modules.

Elix-x committed:
Moved block definitions

Moved block definitions to modules.

JoJoDeveloping committed:
Added definitions for "Crafting"

JoJoDeveloping committed:
Reworked ItemBlock definitions
You can now pass a IItemBlockCustomizer to ease creating a custom ItemBlock instance

JoJoDeveloping committed:
Typo fix

Elix-x commented on src/main/java/appeng/core/crafting/definitions/CraftingTileDefinitions.java:
```diff
@@ -2,19 +2,40 @@
+	private final TileDefinition crafting_monitor;
+	private final TileDefinition crafting_storage;
+	private final TileDefinition crafting_unit;
+	private final TileDefinition molecular_assembler;
```
Uhmmm... No... We said that we kept default java variable naming in code, didn't we?

Elix-x commented on src/api/java/appeng/core/crafting/api/definitions/ICraftingTileDefinitions.java:
```diff
@@ -11,4 +10,21 @@
+        return get( "crafting_unit" );
+    }
+
+    default ITileDefinition tileMolecularAssembler(){
```
We are already in tile definitions, so adding tile/block/item prefix is completely unnecessary.

Elix-x reviewed (commented):
Thanks for help. Overall, it's fine, except for a few names. I guess we weren't clear on naming convention.

Elix-x commented on src/main/java/appeng/core/lib/features/BlockDefinition.java:
```diff
@@ -85,4 +89,13 @@ public boolean isSameAs( Object other )
 		}
 	}
 
+	public IItemDefinition buildItemBlock( FeatureFactory factory){
```
Definitions shouldn't have anything build releated in them. Because they are already defined... If that makes sense :D .

JoJoDeveloping replied:
Storing them in the FeatureFactory would indeed be better. But thay already contain build-related stuff (setItem()) and are not yet complete when they are build() because the ItemBlock is missing.

Elix-x committed:
Removed IItemBlockCustomizer from built definition

Moved IItemBlockCustomizer from built definition to builders & factory.
Adjusted type params of customizer.

Elix-x committed:
Fixed some compile errors + formatting

Fixed some compile errors + formatting.

Elix-x committed:
Moved tile definitions

Moved tile definitions to modules.
